### DISCUSSION TITLE
Is there an easy way to make the processes/workflows in a dsl2 fork, change based on a condition?

### DISCUSSION THREAD

**Discussion by dileep-kishore** [nextflow-io/nextflow#2112] [Q&A] [answered]
Consider this example:

    workflow example_workflow {
        take:
            input_channel
        main:
            input_channel \
            | foo \
            | (bar1 & bar2 & bar3) \
            | mix
    }

What I want to do is to change the processes in the fork based on a condition or variable:

    variable = ["bar1", "bar2"]
    workflow example_workflow {
        take:
            input_channel
        main:
            input_channel \
            | foo \
            // Some condition or function here
            | (bar1 & bar2) \
            | mix
    }

Is there any way other than writing if and switch statements for every possible combination?

**Comment by pditommaso** [comment]
The branch operator may help https://www.nextflow.io/docs/latest/operator.html#branch

**Reply by dileep-kishore** [reply]
I think that I was a little unclear while framing my question. I don&rsquo;t want to split the channels but instead would like to create dynamic workflows based on user input or a parameter value:

    variable = ["bar1", "bar2"]  // This is the user input
    workflow example_workflow {
        take:
            input_channel
        main:
            input_channel \
            | foo \
            | (bar1 & bar2) \  // <- what we want to change based on the value of variables
            | mix
    }

If the variable was changed the workflow needs to change to:

    variable = ["bar1", "bar2", "bar3"]  // Now we want to execute 3 processes
    workflow example_workflow {
        take:
            input_channel
        main:
            input_channel \
            | foo \
            | (bar1 & bar2 & bar3) \  // <- changed from the previous set of 2 processes to 3
            | mix
    }

Here bar1, bar2, etc are processes and I want to take in a list of processes as input from the configuration and dynamically do the forking.

**Reply by dileep-kishore** [reply]
One idea I had was to write multiple workflows using a loop structure and select the appropriate one based on the input. Is there a better way?

**Comment by dileep-kishore** [comment, marked as answer]
I realized that I can use the `when` directive on processes. Eg:

    process bar1 {
    input:
        ...
    when:
        "bar1" in variable
    output:
        ...
    script:
        """
        ...
        """
    }
